File: backend/app/achievements/rules.py
from typing import Dict, Any, List
from sqlalchemy.orm import Session
from uuid import UUID
from app.models.user import User, RoleEnum
from app.models.portfolio_item import PortfolioItem
from app.models.social_link import SocialLink

class RuleEngine:
    """Evaluates business and promoter rules independently."""

    def __init__(self, session: Session):
        self.session = session

    def evaluate_all(self, user: User) -> List[Dict[str, Any]]:
        achievements_earned = []
        if user.role == RoleEnum.BUSINESS:
            achievements_earned.extend(self._evaluate_business_rules(user))
        elif user.role == RoleEnum.PROMOTER:
            achievements_earned.extend(self._evaluate_promoter_rules(user))
            
        achievements_earned.extend(self._evaluate_general_rules(user))
        return achievements_earned

    def _evaluate_general_rules(self, user: User) -> List[Dict[str, Any]]:
        earned = []
        # Profile completion rule
        if user.role == RoleEnum.PROMOTER:
            if hasattr(user, 'promoter_profile') and user.promoter_profile:
                earned.append({"key": "COMPLETE_PROFILE", "progress": 100.0})
        elif user.role == RoleEnum.BUSINESS:
            if hasattr(user, 'business_profile') and user.business_profile:
                earned.append({"key": "COMPLETE_BUSINESS_PROFILE", "progress": 100.0})
                
        # Social links
        social_links = self.session.query(SocialLink).filter(SocialLink.user_id == user.id).count()
        if social_links > 0:
            earned.append({"key": "FIRST_SOCIAL_LINK", "progress": 100.0})
            
        return earned

    def _evaluate_business_rules(self, user: User) -> List[Dict[str, Any]]:
        earned = []
        # Add rules based on existing models here if needed.
        return earned

    def _evaluate_promoter_rules(self, user: User) -> List[Dict[str, Any]]:
        earned = []
        # Check portfolio items
        portfolios = 0
        if hasattr(user, 'promoter_profile') and user.promoter_profile:
            portfolios = self.session.query(PortfolioItem).filter(PortfolioItem.promoter_id == user.promoter_profile.id).count()
        if portfolios > 0:
            earned.append({"key": "FIRST_PORTFOLIO", "progress": 100.0})
        return earned

File: backend/src/modules/achievements/rules.js
import { prisma } from "../../config/db.js";
import { ROLE } from "../../shared/enums.js";

/**
 * Rule engine: evaluates achievement progress for a user.
 * Returns [{ key, progress }] entries (progress 0..100).
 * `user` must include promoterProfile and businessProfile relations.
 */
export async function evaluateAll(user) {
  const earned = [];

  if (user.role === ROLE.PROMOTER) {
    earned.push(...(await evaluatePromoterRules(user)));
  }

  earned.push(...(await evaluateGeneralRules(user)));
  return earned;
}

async function evaluateGeneralRules(user) {
  const earned = [];
  if (user.role === ROLE.PROMOTER && user.promoterProfile) {
    earned.push({ key: "COMPLETE_PROFILE", progress: 100 });
  } else if (user.role === ROLE.BUSINESS && user.businessProfile) {
    earned.push({ key: "COMPLETE_BUSINESS_PROFILE", progress: 100 });
  }

  const socialLinks = await prisma.socialLink.count({ where: { userId: user.id } });
  if (socialLinks > 0) {
    earned.push({ key: "FIRST_SOCIAL_LINK", progress: 100 });
  }
  return earned;
}

async function evaluatePromoterRules(user) {
  const earned = [];
  if (user.promoterProfile) {
    const portfolios = await prisma.portfolioItem.count({
      where: { promoterId: user.promoterProfile.id },
    });
    if (portfolios > 0) {
      earned.push({ key: "FIRST_PORTFOLIO", progress: 100 });
    }
  }
  return earned;
}
